File: examples/governance/compliance_domain/lib/compliance_domain/exemption/commands/request_exemption.rb
module ComplianceDomain
  class Exemption
    module Commands
      class RequestExemption
        include Hecks::Command
        emits "RequestedExemption"

        attr_reader :model_id
        attr_reader :policy_id
        attr_reader :requirement
        attr_reader :reason

        def initialize(
          model_id: nil,
          policy_id: nil,
          requirement: nil,
          reason: nil
        )
          @model_id = model_id
          @policy_id = policy_id
          @requirement = requirement
          @reason = reason
        end

        def call
          Exemption.new(
            model_id: model_id,
            policy_id: policy_id,
            requirement: requirement,
            reason: reason,
            status: "requested"
          )
        end
      end
    end
  end
end

File: examples/governance_static_go/domain/request_exemption.go
package domain

import (
	"time"
)

type RequestExemption struct {
	ModelId string `json:"model_id"`
	PolicyId string `json:"policy_id"`
	Requirement string `json:"requirement"`
	Reason string `json:"reason"`
}

func (c RequestExemption) CommandName() string { return "RequestExemption" }

func (c RequestExemption) Execute(repo ExemptionRepository) (*Exemption, *RequestedExemption, error) {
	agg := NewExemption(c.ModelId, c.PolicyId, c.Requirement, c.Reason, "", time.Time{}, time.Time{}, "", "")
	agg.Status = "requested"
	if err := agg.Validate(); err != nil {
		return nil, nil, err
	}
	if err := repo.Save(agg); err != nil {
		return nil, nil, err
	}
	event := RequestedExemption{
		AggregateID: agg.ID,
		ModelId: c.ModelId,
		PolicyId: c.PolicyId,
		Requirement: c.Requirement,
		Reason: c.Reason,
		OccurredAt: time.Now(),
	}
	return agg, &event, nil
}
